cpu/atmega_common/periph/rtc.c

```diff
@@ -89,15 +86,13 @@ int rtc_set_time(struct tm *time)
 
 int rtc_get_time(struct tm *time)
 {
-    bool tmp = isr_flag;
-
-    *time = tm_now;
+    uint8_t before;
 
-    /* check if interrupt happened while
-       reading the time struct */
-    if (isr_flag != tmp) {
+    /* loop in case of overflow */
+    do {
+        before = TCNT2;
         *time = tm_now;
-    }
+    } while (before > TCNT2);
```

Comment:
```suggestion
    do {
        before = TCNT2;
        /* prevent compiler from reordering memory access to tm_now;
         * including moving it out of the loop */
        __asm__ volatile ("" : : "memory");
        *time = tm_now;
    } while (before > TCNT2);
```

Otherwise the code below would be considered to be a valid optimization of the above code:

```C
    do {
        before = TCNT2;
    } while (before > TCNT2);
    *time = tm_now;
```

Or even the following would be otherwise valid:

```C
    *time = tm_now;
    do {
        before = TCNT2;
    } while (before > TCNT2);
```

Reasoning: Only the reads from the (`volatile` qualified) `TCNT2` are considered to be observable behavior and the value of `tm_now` would be considered to be unchanged (as the C compiler doesn't care about concurrency / IRQs). The memory barrier will force the compiler to assume that all memory (and hence also `tm_now`) has potentially been changed in between and read it again after loading `before`.

Reply:
good catch! I'll apply the same fix in #16347